R/runClusterSummaryMetrics.R

```diff
@@ -0,0 +1,61 @@
+#' @title Run Cluster Summary Metrics
+#' @description Calculates the mean expression of percent of cells that express the
+#' given genes for each cluster
+#'
+#' @param inSCE The single cell experiment to use.
+#' @param useAssay The assay to use.
+#' @param feature A string or vector of strings with each gene to aggregate.
+#' @param displayName A string that is the name of the column used for genes.
+#' @param clusters The name of a colData entry that can be used as groups.
+#' @return A dataframe with mean expression and percent of cells in cluster that 
+#' express for each cluster.
+#' @examples
+#' data("scExample")
+#' runClusterSummaryMetrics(inSCE=sce, useAssay="counts", feature=c("B2M", "MALAT1"), 
+#' displayName="feature_name", clusters="type")
+#' @export
+
+runClusterSummaryMetrics <- function(inSCE, useAssay="logcounts", feature, displayName=NULL, clusters="cluster"){
+  
+  if (!clusters %in% names(SingleCellExperiment::colData(inSCE))) {
+    stop("Specified variable '", clusters, "' not found in colData(inSCE)")
+  }
+  if (!is.null(displayName)) {
+    dimnames(inSCE)[[1]] <- rowData(inSCE)[[displayName]]
+    dimnames(assay(inSCE, i=useAssay, withDimnames=FALSE))[[1]] <- rowData(inSCE)[[displayName]]
+    falseGenes <- setdiff(feature, rowData(inSCE)[[displayName]])
+    feature <- intersect(feature, rowData(inSCE)[[displayName]])
+    warning <- paste0("rowData(inSCE)$", displayName)
+  }
+  else {
+    falseGenes <- setdiff(feature, dimnames(inSCE)[[1]])
+    feature <- intersect(feature, dimnames(inSCE)[[1]])
+    warning <- "dimnames"
+  }
+  
+  if (length(feature) == 0) {
+    stop("All genes in '", toString(falseGenes), "' not found in ", warning)
+  }
+  if (length(falseGenes) > 0) {
+    warning("Specified genes '", toString(falseGenes), "' not found in ", warning)
+  }
+  
+  avgExpr <- data.frame(assay(scuttle::aggregateAcrossCells(inSCE, ids=SingleCellExperiment::colData(inSCE)[,clusters], 
+                                                            statistics="mean", use.assay.type=useAssay, 
+                                                            subset.row=feature)), check.names=FALSE)
+  
+  Gene <- NULL
+  avgExpr$Gene <- row.names(avgExpr)
```

Comment:
Never mind, I answered my own question. It looks like avgExpr and percExpr are dfs. I was thinking these would get changed to a vector. So the final output would be a list of 3 vectors

```
return(list(features=, avgExpr=avgExpr, percExpr=percExpr))
```

where each thing is a single vector. Actually, I would be okay if this were a df instead of a list

```
return(df(features=gene, avgExpr=avgExpr, percExpr=percExpr))
```

And the you would pass these into .ggBubble separately:

```
.ggBubble(features=df$features, percExpr=df$perExpr, avgExpr=df$Expr).
```

Note that I think you may need to add a "features" parameter. The overall goal is just to pass vectors into the lowest function (i.e. .ggBubble) rather than a data frame that requires specific named headers.